source/connection_manager.c

```diff
@@ -0,0 +1,876 @@
+    AWS_ZERO_STRUCT(errors);
+
+    if (new_connections > 0) {
+        AWS_LOGF_INFO(
+            AWS_LS_HTTP_CONNECTION_MANAGER,
+            "id=%p: Requesting %zu new connections from http",
+            (void *)manager,
+            new_connections);
+    }
+
+    /* Even if we can't init this array, we still need to invoke error callbacks properly */
+    bool push_errors =
+        aws_array_list_init_dynamic(&errors, manager->allocator, new_connections, sizeof(int)) == AWS_ERROR_SUCCESS;
+
+    for (size_t i = 0; i < new_connections; ++i) {
+        if (s_aws_http_connection_manager_new_connection(manager)) {
+            ++new_connection_failures;
+            representative_error = aws_last_error();
+            if (push_errors) {
+                AWS_FATAL_ASSERT(aws_array_list_push_back(&errors, &representative_error) == AWS_OP_SUCCESS);
+            }
+        }
+    }
+
+    if (new_connection_failures > 0) {
+        /*
+         * We failed and aren't going to receive a callback, but the current state assumes we will receive
+         * a callback.  So we need to re-lock and update the state ourselves.
+         */
+        aws_mutex_lock(&manager->lock);
+
+        AWS_FATAL_ASSERT(manager->pending_connects_count >= new_connection_failures);
+        manager->pending_connects_count -= new_connection_failures;
+
+        /*
+         * Rather than failing one acquisition for each connection failure, if there's at least one
+         * connection failure, we instead fail all excess acquisitions, since there's no pending
+         * connect that will necessarily resolve them.
+         *
+         * Try to correspond an error with the acquisition failure, but as a fallback just use the
+         * representative error.
+         */
+        size_t i = 0;
+        while (manager->pending_acquisition_count > manager->pending_connects_count) {
+            int error = representative_error;
+            if (i < aws_array_list_length(&errors)) {
+                aws_array_list_get_at(&errors, &error, i);
+            }
+
+            s_aws_http_connection_manager_move_front_acquisition(manager, NULL, error, completions);
+            ++i;
+        }
+
+        aws_mutex_unlock(&manager->lock);
+    }
+
+    s_aws_http_connection_manager_complete_acquisitions(completions, manager->allocator);
+
+    aws_array_list_clean_up(&errors);
+}
+
+int aws_http_connection_manager_acquire_connection(
+    struct aws_http_connection_manager *manager,
+    aws_http_on_client_connection_setup_fn *callback,
+    void *user_data) {
+
+    int result = AWS_OP_ERR;
+
+    struct aws_http_connection_acquisition *request =
+        aws_mem_acquire(manager->allocator, sizeof(struct aws_http_connection_acquisition));
+    if (request == NULL) {
+        callback(NULL, aws_last_error(), user_data);
+        return result;
+    }
+
+    AWS_LOGF_DEBUG(AWS_LS_HTTP_CONNECTION_MANAGER, "id=%p: Acquire connection", (void *)manager);
```

Comment:
TRACE?

Reply:
I'm not sure if I'm out of sync with others here, but I tend to think of trace as debug info that's just too noisy/high frequency to otherwise log at the debug level.  Since this is per connection-acquire request, it feels like it can go in debug.